src/x86_64.rs

```diff
@@ -0,0 +1,331 @@
+///     pub limit: __u16,
+///     pub padding: [__u16; 3usize],
+/// }
+pub use kvm_bindings::kvm_dtable as DescriptorTable;
+
+/// Floating Pointe Unit State
+///
+/// pub struct kvm_fpu {
+///    pub fpr: [[__u8; 16usize]; 8usize],
+///    pub fcw: __u16,
+///    pub fsw: __u16,
+///    pub ftwx: __u8,
+///    pub pad1: __u8,
+///    pub last_opcode: __u16,
+///    pub last_ip: __u64,
+///    pub last_dp: __u64,
+///    pub xmm: [[__u8; 16usize]; 16usize],
+///    pub mxcsr: __u32,
+///    pub pad2: __u32,
+/// }
+pub use kvm_bindings::kvm_fpu as FpuState;
+
+/// Entry describing a CPUID feature/leaf. Features can be set as responses to
+/// the CPUID instruction.
+/// pub struct kvm_cpuid_entry2 {
+///    pub function: __u32,
+///    pub index: __u32,
+///    pub flags: __u32,
+///    pub eax: __u32,
+///    pub ebx: __u32,
+///    pub ecx: __u32,
+///    pub edx: __u32,
+///    pub padding: [__u32; 3usize],
+/// }
+use kvm_bindings::kvm_cpuid_entry2 as CpuIdEntry2;
+
+/// Array of CpuId2 entries, each of which describing a feature/leaf to be set
+/// pub struct kvm_cpuid2 {
+///    pub nent: __u32,
+///    pub padding: __u32,
+///    pub entries: __IncompleteArrayField<kvm_cpuid_entry2>,
+/// }
+use kvm_bindings::kvm_cpuid2 as CpuId2;
+
+/// Unix definition of the LAPIC state, the set of memory mapped registers that
+/// describe the Local APIC. Unix-based VMMs only require 1KB of memory to
+/// describe the LAPIC state.
+///
+/// pub struct kvm_lapic_state {
+///    pub regs: [::std::os::raw::c_char; 1024usize],
+/// }
+#[cfg(unix)]
+pub use kvm_bindings::kvm_lapic_state as LapicState;
+
+/// Windows definition of the LAPIC state, the set of memory mapped registers
+/// that describe the Local APIC. Windows-based VMMs require 4KB of memory to
+/// describe the LAPIC state, or the Windows APIs will fail, even though the
+/// architecture-specified space requirement is only 1KB.
+#[cfg(windows)]
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct LapicState {
+    pub regs: [::std::os::raw::c_char; 4096usize],
+}
+
+#[cfg(windows)]
+impl Default for LapicState {
+    fn default() -> Self {
+        unsafe { ::std::mem::zeroed() }
+    }
+}
+
+#[cfg(windows)]
+impl ::std::fmt::Debug for LapicState {
+    fn fmt(&self, fmt: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        self.regs[..].fmt(fmt)
+    }
+}
+
+#[test]
+fn vcpu_test_layout_lapic_state() {
+    assert_eq!(
+        ::std::mem::size_of::<LapicState>(),
+        1024usize,
+        concat!("Size of: ", stringify!(LapicState))
+    );
+    assert_eq!(
+        ::std::mem::align_of::<LapicState>(),
+        1usize,
+        concat!("Alignment of ", stringify!(LapicState))
+    );
+    assert_eq!(
+        unsafe { &(*(::std::ptr::null::<LapicState>())).regs as *const _ as usize },
+        0usize,
+        concat!(
+            "Offset of field: ",
+            stringify!(LapicState),
+            "::",
+            stringify!(regs)
+        )
+    );
+}
+
```

Comment:
@andreeaflorescu I pulled pulled these data structure definitions from kvm-ioctls/mod.rs (ie, CpuId wrapper), as these are not kvm-specific and are useful to any Vcpu implementation. The duplicate will be removed in the next-step PR to update `kvm-ioctls` to use this `vmm-vcpu` crate. Let me know if you prefer another way.

Reply:
We are currently working on providing a more flexible interface for kvm structures with 0-sized arrays. Can you also take a look at this PR: https://github.com/rust-vmm/kvm-ioctls/pull/6 and let us know what you think?

We need to alter cpuid entries (adding new ones to the list, removing existing ones and so on). This is the main purpose of this PR.

Reply:
This is nice. At first glance it looks good, but I'll also take a deeper look. It also resolves a question I had, in whether the MSR array was going to subject to the same issues as the CPUID array (looks like both are resolved in this PR).

Reply:
Yap, it is a generic solution for handling 0-sized arrays.